### PR TITLE
feat: Shadow DOM

### PR DESCRIPTION
**Pull request by Marshal27**
@devknoll Created this PR based on your #53  

impl containing the following:
- monkeypatch `CSSStyleSheet.replace` and `CSSStyleSheet.replaceSync`.
- associate the container descriptors with the CSSStyleSheet on shadowRoot.
- pass the set of query descriptors as part of the state/context.
- only monkeypatch `CSSStyleSheet.replace` and `CSSStyleSheet.replaceSync` if prototype function is defined.

pending:
- need to test/and or possibly implement support for `::slotted()` and `::part()`
- need to complete further testing with `:host()` (so far base level testing here of setting `container-type` seems to be working).

### CODE REVIEW

**Comment by Marshal27** [comment]
Just realized you pushed a commit with some significant changes while I was working on this. 

I will take a look at the new iteration.

File: src/engine.ts [hunk 1]
```diff
@@ -210,6 +210,40 @@ export function initializePolyfill() {
     attributeOldValue: true,
   });
 
+  const originalAttachShadow = Element.prototype.attachShadow;
+  Element.prototype.attachShadow = function (options) {
+    const shadow = originalAttachShadow.apply(this, [options]);
+    mutationObserver.observe(shadow, {
```
Comment (devknoll) [inline src/engine.ts:216]: We probably want to move the `observe` to the `connect` handler of the controller, so we can `unobserve` during disconnect. Inside `attachShadow`, it should just be sufficient to call `getOrCreateInstance(shadow)`

File: src/engine.ts [hunk 2]
```diff
@@ -210,6 +210,40 @@ export function initializePolyfill() {
...
+      childList: true,
+      subtree: true,
+      attributes: true,
+      attributeOldValue: true,
+    });
+    return shadow;
+  }
+
+  const originalReplaceSync = CSSStyleSheet.prototype.replaceSync;
```
Comment (devknoll) [inline src/engine.ts:225]: I think it would be sufficient to just not monkeypatch these if they aren't already defined.

File: src/engine.ts [hunk 3]
```diff
@@ -399,6 +435,18 @@ export function initializePolyfill() {
       const maybeComputeState: (
         parentState: LayoutState
       ) => [LayoutState, boolean] = parentState => {
+        if (node instanceof ShadowRoot) {
```
Comment (devknoll) [inline src/engine.ts:438]: We shouldn't mutate the `parentState`. Instead, we should apply the conditions while computing the state for the `ShadowRoot`.

For now, we can do that by passing a `parentState => QueryDescriptorEntry[]` function to `createStateComputer`. For `ShadowRoot`, that will use the `ShadowRoot.styleSheets` and `ShadowRoot.adoptedStyleSheets` (if present) lists.

File: src/engine.ts [hunk 4]
```diff
@@ -399,6 +435,18 @@ export function initializePolyfill() {
...
+          const adoptedStyleSheet = node.adoptedStyleSheets[0];
```
Comment (devknoll) [inline src/engine.ts:439]: We'll want to do this for every stylesheet, not just the first

File: src/engine.ts [hunk 5]
```diff
@@ -508,6 +556,10 @@ export function initializePolyfill() {
           }
 
           for (const child of node.childNodes) {
+            const element = child as Element;
+            if (element?.shadowRoot) {
```
Comment (devknoll) [inline src/engine.ts:560]: `shadowRoot` won't be available for closed shadow roots. Instead, we should add a `CQ_SHADOW` symbol that we can use to associate the instance with the element (similar to how we associate instances with nodes, and descriptors with stylesheets).

**Comment by devknoll** [comment]
Apologies for the churn -- I was hoping to get those changes in _before_ you rebased 🙂

**Comment by Marshal27** [comment]
> Apologies for the churn -- I was hoping to get those changes in _before_ you rebased 🙂

no problem at all, I welcome the feedback/interaction 😄

File: src/engine.ts (hunk 1 above)
Comment (Marshal27) [reply src/engine.ts:216]: you mentioned `unobserve` in the `controller.disconnect`, should this be changed to the `ResizeObserver` or call `MutationObserver.disconnect` in the `controller.disconnect`?

Comment (devknoll) [reply src/engine.ts:216]: `MutationObserver` -- e.g. if the tree containing the `ShadowRoot` is disconnected, we want to clean it up as part of the disconnect.

In order to expose the `MutationObserver` to the `ShadowRootController`, I would recommend something like:

```ts
interface Disposable {
  (): void;
}

interface MutationObserverContext {
  observeMutation(node: Node): Disposable;
}
```

After the `MutationObserver` is created, just declare:

```
function observeMutation(node: Node) {
  mutationObserver.observe(node);
  return () => {
    mutationObserver.unobserve(node);
  };
}
```

Then pass this to the `ShadowRootController` similar to how other contexts are being passed. Then you can observe in `connected` and dispose the object in `disconnected` to clean up.

Comment (Marshal27) [reply src/engine.ts:216]: Thank you for the context -- the implementation example is very helpful!

I am familiar with the pattern and why we need to implement it, but not as familiar with the nature of the `MutationObserver.disconnect` function. My curiosity was primarily around if `disconnect` would stop all mutation observations, or, if it would only stop the observation on the node the invocation is scoped in. (for example being able to target nodes with `ResizeObserver.unobserve(node)`
- I wasn't sure if there is a situation of only needing to stop observation on the `ShadowRoot` specifically.
- Or maybe in this particular implementation, it is ok to call `disconnect` during that life-cycle as we want to stop all observations and `observe` will be called again from the top down etc?

Comment (devknoll) [reply src/engine.ts:216]: D'oh. My bad. Of course `MutationObserver` doesn't have an `uobserve` function :) We'll want to create a new one for each `ShadowRoot` then (and reuse the same callback), where it'll be safe to call `disconnect()`.

File: src/engine.ts (hunk 3 above)
Comment (Marshal27) [reply src/engine.ts:438]: I have all of the other change requests done, but this is the last piece. I am struggling to figure out how to pass the styles from the `ShadowRoot.adoptedStyleSheets` into `createStateComputer` (since it uses `window.getComputedStyle(element)` which doesn't work for `asoptedStyleSheets`)... could you send me some sample syntax on implementing this bit?

Comment (devknoll) [reply src/engine.ts]: We don't need to pass the style information to `createStateComputer` -- we just need the list of stylesheets, so that we can call `getDescriptorsForStyleSheet` on them.

```js
interface QueryDescriptorProvider {
  (parentState: LayoutState): ContainerConditionEntry[];
}

function createStateComputer(el: Element, queryDescriptorProvider: QueryDescriptorProvider) {
  /* ... */
}
```

When creating the `createStateComputer` for the Shadow DOM, we would pass one in that reads from the adopted styles. For everything else, we would just read from `parentState`.

File: src/types/global.d.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+export {};
+
+declare global {
+    interface ShadowRoot {
+        adoptedStyleSheets: CSSStyleSheet[];
```
Comment (devknoll) [inline src/types/global.d.ts:5]: nit: we should use the same type as `document.styleSheets` here. 

I think we could also move this to the existing `globals.d.ts` file?

File: src/engine.ts [hunk 7]
```diff
@@ -210,13 +211,47 @@ export function initializePolyfill() {
...
+        options,
+        undefined
+      );
+      setDescriptorsForStyleSheet(this, result.descriptors);
+      const replaceSync = originalReplaceSync.apply(this, [result.source]);
+      return replaceSync;
+    }
+  }
+
+  const originalReplace = CSSStyleSheet.prototype.replace;
```
Comment (devknoll) [inline src/engine.ts:242]: For completeness, I think we would also want to patch `insertRule` and `deleteRule`. For now though, I think you could just include a TODO comment about how they aren't handled yet.

File: src/engine.ts [hunk 8]
```diff
@@ -381,6 +416,8 @@ export function initializePolyfill() {
               ...options,
             }),
         });
+      } else if (node instanceof ShadowRoot) {
+        innerController = new ShadowStyleSheetController(node, createMutationObserver());
```
Comment (devknoll) [inline src/engine.ts:420]: You can set `stateProvider` here instead of having the separate check later.

**Comment by devknoll** [comment]
#53 was merged, so you should be able to rebase against main now.

**Comment by Marshal27** [comment]
@devknoll if you are good with current impls, would you be ok with me rebasing and landing this on `main` and then handling any additional work for `::slotted()` `::part()` or `:host()` in a separate PR?

**Comment by Marshal27** [comment]
@devknoll this is ready for final review/merge when you have a moment to do another pass.

**Comment by devknoll** [comment]
Very sorry for the delay, I've been swamped with other things. Will review this today or tomorrow!

File: src/engine.ts [hunk 9]
```diff
@@ -644,6 +701,70 @@ class StyleElementController extends NodeController<HTMLStyleElement> {
   }
 }
 
+class ShadowRootController extends NodeController<ShadowRoot> {
+  private controller: AbortController | null = null;
+  private mo: MutationObserver | null = null;
+  private host: HTMLElement | null = null;
```
Comment (devknoll) [inline src/engine.ts:707]: These don't seem like they need the `null` case?

**Review by devknoll — COMMENTED** [review]
> if you are good with current impls, would you be ok with me rebasing and landing this on main and then handling any additional work for ::slotted() ::part() or :host() in a separate PR?

@Marshal27 I'm not too keen on landing partial feature support. That said, I think enabling shadow roots _at all_ may be a good incremental step, so I think I would be willing to land this if we deferred the `:host` and related changes (e.g. `processCssRules`), so that this PR was focused on just allowing you to use containers inside a shadow root that were defined inside that shadow root. The `processCssRules` change is going to need some further iteration anyway I think. 

How does that sound?

**Comment by Marshal27** [comment]
> > if you are good with current impls, would you be ok with me rebasing and landing this on main and then handling any additional work for ::slotted() ::part() or :host() in a separate PR?
> 
> @Marshal27 I'm not too keen on landing partial feature support. That said, I think enabling shadow roots _at all_ may be a good incremental step, so I think I would be willing to land this if we deferred the `:host` and related changes (e.g. `processCssRules`), so that this PR was focused on just allowing you to use containers inside a shadow root that were defined inside that shadow root. The `processCssRules` change is going to need some further iteration anyway I think.
> 
> How does that sound?

I agree with you on not landing partial work, but, I also agree with your recommended approach to keep this focused on enabling containers in the `shadowRoot` etc. Based on recent dialog (#56 Support for non-trivial CSS Selectors) the `processCssRules` work will likely be driven by patterns adopted for that topic so it makes sense in my opinion to compartmentalize the `processCssRules` work.

File: src/engine.ts [hunk 10]
```diff
@@ -813,113 +934,136 @@ function computeDimensionSum(
   return names.reduce((value, name) => value + computeDimension(read, name), 0);
 }
 
-function createStateComputer(element: Element) {
-  const styles = window.getComputedStyle(element);
-
+function createStateComputer(shadow: ShadowRoot): (parentState: LayoutState, cacheKey: symbol) => LayoutState;
+function createStateComputer(element: Element): (parentState: LayoutState, cacheKey: symbol) => LayoutState;
+function createStateComputer(shadowOrElement: ShadowRoot | Element) {
+  const isElement = shadowOrElement instanceof Element;
```
[the remark this answers is not in the document]
Comment (Marshal27) [reply src/engine.ts:940]: no problem, I wasn't sure about this approach either... I will get this updated.

File: src/engine.ts (hunk 9 above)
Comment (Marshal27) [reply src/engine.ts:707]: will get this updated

File: src/engine.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Marshal27) [reply src/engine.ts:940]: @devknoll I am struggling with this a bit. My original implementation was to mutate the parent state here.
https://github.com/GoogleChromeLabs/container-query-polyfill/pull/57#discussion_r983677469_

can you give me an example of what this shadow root specific state update function may look like?

[the remark this answers is not in the document]
Comment (Marshal27) [reply src/engine.ts]: @devknoll let me know if what I just pushed is what you had in mind.

[the remark this answers is not in the document]
Comment (devknoll) [reply src/engine.ts]: The states should all be immutable. The ShadowRoot would clone the parent state and overwrite the field. Your current version is pretty close, I'll leave a comment inline.

File: src/engine.ts [hunk 11]
```diff
@@ -416,6 +456,26 @@ export function initializePolyfill() {
...
+            for (const query of getDescriptorsForStyleSheet(adoptedStyleSheet)) {
+              rootQueryDescriptors.push([
+                new Reference(query),
+                QueryContainerFlags.None,
+              ]);
+            }
+          }
+        }
+        if (rootQueryDescriptors.length) {
+          state.conditions = rootQueryDescriptors;
```
Comment (devknoll) [inline src/engine.ts:475]: The parent state is immutable. So this should be:

```ts
return {
  ...state,
  conditions: rootQueryDescriptors,
};
```

File: src/engine.ts [hunk 12]
```diff
@@ -517,6 +577,12 @@ export function initializePolyfill() {
           for (const child of node.childNodes) {
             getOrCreateInstance(child).update(currentState);
           }
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const shadow = (node as any)[SHADOW_SYMBOL];
+          if (shadow) {
```
Comment (devknoll) [inline src/engine.ts:582]: We should just set this as the `stateProvider` at the same time we set the `innerController`

File: src/engine.ts [hunk 13]
```diff
@@ -360,7 +398,7 @@ export function initializePolyfill() {
   function getOrCreateInstance(node: Node): Instance {
     let instance = getInstance(node);
     if (!instance) {
-      let innerController: NodeController<Node>;
+      let innerController: NodeController<Node|ShadowRoot>;
```
Comment (devknoll) [inline src/engine.ts:401]: `ShadowRoot` should inherit from `Node` so this shouldn't be necessary anymore

File: src/engine.ts [hunk 14]
```diff
@@ -644,6 +720,68 @@ class StyleElementController extends NodeController<HTMLStyleElement> {
...
+
+  connected(): void {
+    this.mo?.observe(this.node, {
+      childList: true,
+      subtree: true,
+      attributes: true,
+      attributeOldValue: true,
+    });
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (this.host as any)[SHADOW_SYMBOL] = this.node;
```
Comment (devknoll) [inline src/engine.ts:742]: I think we should probably just set this in `attachShadow`.

File: src/engine.ts (hunk 12 above)
Comment (Marshal27) [reply src/engine.ts:582]: ahhh! this was the piece I was struggling to see, this makes sense.

File: src/engine.ts (hunk 13 above)
Comment (Marshal27) [reply src/engine.ts:401]: good call out, I wasn't sure about that bit when I added it.

File: src/engine.ts (hunk 14 above)
Comment (Marshal27) [reply src/engine.ts:742]: great call out here, that definitely makes more sense.

**Comment by Marshal27** [comment]
@devknoll if you spot any additional items, let me know, and I will get them implemented... if not, should be GTG.

File: src/engine.ts [hunk 15]
```diff
@@ -417,6 +460,27 @@ export function initializePolyfill(updateCallback: () => void) {
         ];
       };
 
+      const updateShadowHostState: (
+        node: ShadowRoot,
+        state: LayoutState
+      ) => LayoutState = (node, state) => {
```
Comment (devknoll) [inline src/engine.ts:466]: You could make this `createShadowHostUpdater` that takes the `ShadowRoot` and returns a `LayoutStateProvider`, and move it to the top level scope.

File: src/engine.ts [hunk 16]
```diff
@@ -529,6 +593,12 @@ export function initializePolyfill(updateCallback: () => void) {
           ) {
             getOrCreateInstance(child).update(currentState);
           }
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const shadow = (node as any)[SHADOW_SYMBOL];
+          if (shadow) {
+            updateShadowHostState(shadow, currentState);
```
Comment (devknoll) [inline src/engine.ts:599]: This shouldn't be needed here anymore: the state updater should be called by the `update` call

File: src/engine.ts [hunk 17]
```diff
@@ -661,6 +741,66 @@ class StyleElementController extends NodeController<HTMLStyleElement> {
...
+  }
+
+  private processCssRules(): void {
+    let setProp = false;
+    for (const adoptedStyleSheet of this.node.adoptedStyleSheets) {
+      for (const rule of adoptedStyleSheet.cssRules) {
+        if (rule instanceof CSSStyleRule) {
+          const value = rule.style.getPropertyValue(CUSTOM_PROPERTY_TYPE).trim();
+          if (value) {
+            const selector: string = rule.selectorText.replace(':host', this.host.localName);
```
Comment (devknoll) [inline src/engine.ts:785]: I still think I would prefer to have this moved into a separate PR. Without these changes, you should be able to use container queries in a shadow DOM, just without supporting the shadow DOM selectors.

Then we can work on finalizing support for those selectors

File: src/engine.ts (hunk 16 above)
Comment (Marshal27) [reply src/engine.ts:599]: Great catch! I forgot to drop this when moving the function to assign the `stateProvider`.

File: src/engine.ts (hunk 17 above)
Comment (Marshal27) [reply src/engine.ts:785]: Sorry for my confusion on this part, I didn't realize you were wanting to move this out completely. 

Because we were discussing making this PR the base impl to enable `container-type` functionality from within the shadow DOM, I viewed this as being part of the impl, as it would not be uncommon for developers to want to apply `container-type` to the `host` from within the shadow DOM via the `:host` selector. 
- This is a minimum impl to support `container-type` on the `:host` selector from within the shadow DOM, re-process the rules on mutation events incase attributes are changed on the host, and nothing more.

```css 
:host { container-type: inline-size; }
``` 
or  
```css
:host[attribute] { container-type: inline-size; }
```

Do you still want me to move this out?

File: src/engine.ts (hunk 15 above)
Comment (devknoll) [reply src/engine.ts:466]: It's mostly a preference: I tend to prefer that functions describe what they do with a verb when possible.

`updateShadowHostState` is different from `defaultStateProvider`: `defaultStateProvider` _is_ just a state provider _value_ (i.e. a function that takes a `LayoutState` and returns a new `LayoutState`) that we can use directly. Refactoring `updateShadowHostState` to follow the same pattern as `createStateUpdater` would give you:

```ts
function updateShadowHostState(node: ShadowRoot) {
  return (state: LayoutState) => {
    const rootQueryDescriptors: ContainerConditionEntry[] = [];
    for (const adoptedStyleSheet of node.adoptedStyleSheets) {
      if (adoptedStyleSheet) {
        for (const query of getDescriptorsForStyleSheet(adoptedStyleSheet)) {
          rootQueryDescriptors.push([
            new Reference(query),
            QueryContainerFlags.None,
          ]);
        }
      }
    }
    return {
      ...state,
      conditions: rootQueryDescriptors,
    };
  };
}
```

Now, the name `updateShadowHostState` doesn't make much sense anymore, because it's no longer describing what the function does, which is return a new `LayoutStateProvider` given a `ShadowRoot` element, hence the suggestion that it's renamed to `createShadowHostUpdater` (though I now believe `createShadowRootStateProvider` would be a better name).

Hope that helps!

File: src/globals.d.ts [hunk 18]
```diff
@@ -13,3 +13,10 @@
 
 declare const IS_WPT_BUILD: boolean;
 declare const PACKAGE_VERSION: string;
+interface ShadowRoot {
+    adoptedStyleSheets: CSSStyleSheet[];
+}
+interface StyleSheet {
+    replace(text: string): Promise<CSSStyleSheet>;
+    replaceSync(text: string): void;
+}
```
Comment (devknoll) [inline src/globals.d.ts:22]: Are these changes necessary? My `lib.dom.d.ts` declares them already. Maybe we just need to specify a minimum typescript version or something?

File: src/engine.ts (hunk 17 above)
Comment (devknoll) [reply src/engine.ts:785]: Taking a look now...

File: src/engine.ts (hunk 15 above)
Comment (Marshal27) [reply src/engine.ts:466]: implemented, thank you for sending this impl.

File: src/globals.d.ts (hunk 18 above)
Comment (Marshal27) [reply src/globals.d.ts:22]: I get the following without them.

`TS2339: Property 'replace' does not exist on type 'CSSStyleSheet'.`
`TS2339: Property 'replaceSync' does not exist on type 'CSSStyleSheet'.`

For the `Stylesheet`, I believe the issue is that `lib.dom.d.ts` does indeed have them on the extended interface `CSSStyleSheet`, however, not on the base/underlying `StyleSheet` interface.

similar scenario [Property 'addRule' and 'insertRule' does not exist on type 'StyleSheet'](https://stackoverflow.com/questions/44897499/property-addrule-and-insertrule-does-not-exist-on-type-stylesheet)

File: src/engine.ts (hunk 17 above)
Comment (Marshal27) [reply src/engine.ts:783]: @devknoll let me know any thoughts or concerns when you have time to circle back to this.

**Comment by gkatsev** [comment]
Thanks for your work on this! Though, I'm wondering is this supposed to be with just a standard style element or only with Constructable Stylesheets?

I'm trying it out [on a project](https://github.com/muxinc/media-chrome/pull/347) where our existing web component templates include a style element inside of them, and with this PR it didn't work. Though, once I switched it to use constructable stylesheets it worked like a champ.

**Comment by Marshal27** [comment]
@gkatsev I appreciate you bringing that to my attention. The latest push should provide basic support for standard style elements; please give it a test drive and let me know how it goes.

**Please note:**

Attempting either of the following is not currently working for standard style elements, however, these two approaches are currently working for the constructed stylesheets. 
>Because that impl is currently under review by @devknoll I did not want to incorporate it here until I get his thoughts on the approach. Other than that, basic functionality should now be available.

`:host { container-type: inline-size; }`

or

`:host[attribute] { container-type: inline-size; }`

**Comment by gkatsev** [comment]
Awesome, let me try it out!
I don't currently use `:host` with container queries, so, it's working out.

**Comment by gkatsev** [comment]
Seems to be working now without requiring us to adopt Constructable Stylesheets immediately. Thanks!

**Comment by gkatsev** [comment]
Thank you for the PR and for the maintainers for maintaining this polyfill. I hope this gets pulled in soon.

File: src/engine.ts (hunk 17 above)
Comment (Marshal27) [reply src/engine.ts:793]: @devknoll after taking a closer look at this, maybe this is violating `CSS Scoping` and it is ok for the `:host` selector to not support defining a `container-type`?

Comment (Marshal27) [reply src/engine.ts]: Stashed this impl in case it's useful in the future. Went ahead and removed it per your original request.

**Comment by devknoll** [comment]
@Marshal27 Thank you for all of your hard work on this PR. This is an important feature that many folks are passionate about, and one that I had hoped we'd be able to get in.

Unfortunately, I have moved on to other projects and there isn't currently anyone working full-time on the polyfill right now. Without an owner who can help see this full feature through to completion, we don't feel very comfortable merging this PR (or any other new features) right this moment. I'm very sorry about that ☹️

That said, I highly encourage you to continue development on your fork: there are clearly many folks that want this feature and you're best poised to provide it. For testing, the repository is configured to use BrowserStack. You should know that BrowserStack [provides free sponsorship for OSS projects](https://www.browserstack.com/open-source), and we encourage you to apply so that you can continue to test your changes against the official [Web Platform Tests](https://web-platform-tests.org/), which include tests for ShadowDOM support.

My sincere apologies ☹️

**Comment by Marshal27** [comment]
Posting this link here for posterity.

https://github.com/Marshal27/shadow-container-query-polyfill

**Comment by gkatsev** [comment]
Awesome, I was wondering whether you'd be willing to maintain a community fork @Marshal27.

**Comment by Marshal27** [comment]
> Awesome, I was wondering whether you'd be willing to maintain a community fork @Marshal27.

I will definitely do my best. Feel free to add any shadow dom specific issues you may run into over there and I will do what I can.
